### PR TITLE
Dedumbify the IOFile API

### PR DESCRIPTION
**Pull request by delroth**
IOFile is exposing a ton of implementation details through its API. Most importantly, it directly exposes FILE* objects. This led to abuse throughout the codebase, where instead of using the generic IOFile API, users just get a handle from an IOFile and do random shit with it.

This PR cleans up all of this in preparation for removing the use of stdio on Windows (see issue 9273).

Next step will be to clean up the 23 uses of the equally dumb File::OpenFStream.

Note: this is completely untested. I have almost no doubt it will break stuff. So consider it as WIP until it has had sufficient testing.

Things that require more extensive testing:

- [x] PNG writing
- [ ] PPC symboldb

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3524)

### CODE REVIEW

File: Source/Core/Core/HW/DSPLLE/DSPHost.cpp [hunk 1]
```diff
@@ -64,31 +64,10 @@ void CodeLoaded(const u8 *ptr, int size)
...
-		case 0x347112BA: success = DSPSymbols::ReadAnnotatedAssembly("../../docs/DSP/DSP_UC_AXWii.txt"); break;
-		case 0xD643001F: success = DSPSymbols::ReadAnnotatedAssembly("../../docs/DSP/DSP_UC_SuperMarioGalaxy.txt"); break;
-		default: success = false; break;
-	}
-
-	if (!success)
-	{
-		DSPSymbols::AutoDisassembly(0x0, 0x1000);
-	}
+  DSPSymbols::AutoDisassembly(0x0, 0x1000);
```
Comment (lioncash) [inline Source/Core/Core/HW/DSPLLE/DSPHost.cpp]: This looks a little off

File: Source/Core/DiscIO/FileBlob.h [hunk 2]
```diff
@@ -26,9 +27,9 @@ class PlainFileReader : public IBlobReader
 	bool Read(u64 offset, u64 nbytes, u8* out_ptr) override;
 
 private:
-	PlainFileReader(std::FILE* file);
+	PlainFileReader(File::IOFile&& file);
 
-	File::IOFile m_file;
+  File::IOFile m_file;
```
Comment (lioncash) [inline Source/Core/DiscIO/FileBlob.h]: Here too

File: Source/Core/Common/FileUtil.h [hunk 3]
```diff
@@ -159,9 +156,17 @@ bool ReadFileToString(const std::string& filename, std::string& str);
 class IOFile : public NonCopyable
 {
 public:
+	// Flags that can be provided to the Open method or to the IOFile constructor
+	// to get alternative behaviors.
+	enum OpenFlags
```
Comment (BhaaLseN) [inline Source/Core/Common/FileUtil.h]: `enum class`?

File: Source/Core/Common/JitRegister.cpp [hunk 4]
```diff
@@ -48,10 +47,7 @@ void Init(const std::string& perf_dir)
 	{
 		std::string dir = perf_dir.empty() ? "/tmp" : perf_dir;
 		std::string filename = StringFromFormat("%s/perf-%d.map", dir.data(), getpid());
-		s_perf_map_file.Open(filename, "w");
-		// Disable buffering in order to avoid missing some mappings
```
Comment (BhaaLseN) [inline Source/Core/Common/JitRegister.cpp]: Would it hurt to keep the comment, so we know why buffering is disabled?

File: Source/Core/Common/SDCardUtil.cpp [hunk 5]
```diff
@@ -187,12 +187,12 @@ static unsigned int write_empty(FILE* file, u64 count)
 		if (len > count)
 			len = count;
 
-		if ( fwrite(empty, 1, (size_t)len, file) != (size_t)len )
-			return 1;
+		if (!file->WriteBytes(empty, (size_t)len))
```
Comment (BhaaLseN) [inline Source/Core/Common/SDCardUtil.cpp]: How useful is casting `len` to `size_t`, when it should technically be a `size_t` already (except that we store it as `u64` above)? Also, would it make sense to use `ArraySize` or even `std::array` there; while we're at it?

Altho, both those points are probably cleanup and don't belong in this PR...

File: Source/Core/Common/SDCardUtil.cpp [hunk 6]
```diff
@@ -171,12 +171,12 @@ static void fat_init(u8* fat)
 }
 
 
-static unsigned int write_sector(FILE* file, u8* sector)
+static bool WriteSector(File::IOFile* file, u8* sector)
```
Comment (BhaaLseN) [inline Source/Core/Common/SDCardUtil.cpp]: Is there any particular reason to use `IOFile*` here over `IOFile&`? IIRC we opted to use the explicit `func(&foo)` syntax (with `R func(P* parameter)`) on caller site to indicate the value is modified rather than making it a reference (which is not distinguishable from a copy). But in this case we don't intend to change it, so would the opposite reasoning make sense?

Comment (delroth) [inline Source/Core/Common/SDCardUtil.cpp]: It's not const, so we are modifying it.

Comment (BhaaLseN) [inline Source/Core/Common/SDCardUtil.cpp]: Eep, fair point, I keep forgetting about internal state here >_<

**Comment by delroth** [comment]
Dropping for now, let's just hope that MS has their shit fixed. No time for me to work on it.
